mlflow/java/client/src/main/java/org/mlflow/artifacts/CliBasedArtifactRepository.java

```diff
@@ -0,0 +1,242 @@
+    // deserialize a list-of-dictionaries, and then reserialize each dictionary to pass it to
+    // the protobuf deserializer.
+    Gson gson = new Gson();
+    Type type = new TypeToken<List<Map<String, Object>>>(){}.getType();
+    List<Map<String, Object>> listOfDicts = gson.fromJson(json, type);
+    List<Service.FileInfo> fileInfos = new ArrayList<>();
+    for (Map<String, Object> dict: listOfDicts) {
+      String fileInfoJson = gson.toJson(dict);
+      try {
+        Service.FileInfo.Builder builder = Service.FileInfo.newBuilder();
+        JsonFormat.parser().merge(fileInfoJson, builder);
+        fileInfos.add(builder.build());
+      } catch (InvalidProtocolBufferException e) {
+        throw new MlflowClientException("Failed to deserialize JSON into FileInfo: " + json, e);
+      }
+    }
+    return fileInfos;
+  }
+
+  /**
+   * Checks whether the 'mlflow' executable is available, and throws a nice error if not.
+   * If this method has ever run successfully before (in the entire JVM), we will not rerun it.
+   */
+  private void checkMlflowAccessible() {
+    if (mlflowSuccessfullyLoaded.get()) {
+      return;
+    }
+
+    try {
+      String tag = "get mlflow version";
+      String mlflowVersion = forkProcess(Lists.newArrayList(mlflowExecutable, "--version"), tag);
+      logger.info("Found local mlflow executable with version=" + mlflowVersion);
+      mlflowSuccessfullyLoaded.set(true);
+    } catch (MlflowClientException e) {
+      String errorMessage = String.format("Failed to exec process %s, needed to access artifacts " +
+          "within the non-Java-native artifact store at '%s'. Please make sure mlflow is " +
+          "available on your local system path (e.g.," + "from 'pip install mlflow')",
+        mlflowExecutable, artifactBaseDir);
+      throw new MlflowClientException(errorMessage, e);
+    }
+  }
+
+  /**
+   * Forks the given mlflow command and awaits for its successful completion.
+   *
+   * @param command Command used to fork the process.
+   * @param tag User-facing tag which will be used to identify what we were trying to do
+   *            in the case of a failure.
+   * @return raw stdout of the process, decoded as a utf-8 string
+   * @throws MlflowClientException if the process exits with a non-zero exit code, or anything
+   *                               else goes wrong.
+   */
+  private String forkProcess(List<String> command, String tag) {
+    String stdout;
+    Process process = null;
+    try {
+      ProcessBuilder pb = new ProcessBuilder(command);
+      // TODO(aaron) Figure out a way to pass the other fields of the host-creds.
+      pb.environment().put("MLFLOW_TRACKING_URI", hostCredsProvider.getHostCreds().getHost());
+      process = pb.start();
+      stdout = IOUtils.toString(process.getInputStream(), StandardCharsets.UTF_8);
+      int exitValue = process.waitFor();
+      if (exitValue != 0) {
+        throw new MlflowClientException("Failed to " + tag + ". Error: " +
```

Comment:
Since this is asynchronous, can we have built in retries. Esp for server busy or not found error codes? This can be an optional retry counter, for instance. This is not a blocking comment, something to consider.

Reply:
Shouldn't the retries be within `mlflow download`, etc? This method is just forking a process, and that process should probably take care to work well.

Reply:
makes sense